### DISCUSSION TITLE
Jamulus with 32 or lower frame size instead of just 64,128 & 256

### DISCUSSION THREAD

**Discussion by Meemansa-Mishra** [jamulussoftware/jamulus#2844] [Technical Support & Help]
I am trying to run Jamulus with a decreased frame size to acheive lower latency but have been unable to do this.I would appreciate any kind of support in trying to get it down to 32 or less.

**Comment by pljones** [comment]
The protocol doesn't support that.  You'd need to release new versions of the client and server that could translate between the existing frame size and the new frame size, like it currently does for 64 <-> 128 samples.  And, similarly, only when both client and server had the option enabled, would you get end to end support.

**Reply by ghost** [reply]
There are diminishing returns for reducing the buffer size:
- Most operating systems use a 1ms (1KHz) kernel interrupt for scheduling tasks. (48KHz * 1ms = 48 samples in 1ms)
- The kernel interrupt applies to routers also.
- Network data packets are queued to go in efficient bursts.

Note that sound travels about 1 foot in 1 millisecond (less than the distance from your instrument to your ears or brain).

**Reply by Meemansa-Mishra** [reply]
> The protocol doesn't support that. You'd need to release new versions of the client and server that could translate between the existing frame size and the new frame size, like it currently does for 64 <-> 128 samples. And, similarly, only when both client and server had the option enabled, would you get end to end support.

Can you help me with what can I do to achieve this?
Or any suggestions to achieve the lowest possible latency?

**Reply by ann0see** [reply]
> achieve the lowest possible latency

* Use a low latency audio interface
* Use Linux with low latency kernel
* Use Fibre internet only
* Rent a Server close to all musicians with low jitter and ping
* Ban any other traffic from your home network (disable WiFi, unplug LAN) 

I think that's more important  to have a good network than tweaking buffer sizes

**Reply by Meemansa-Mishra** [reply]
@ann0see 
Thanks for your quick responses.

We have already done what you have mentioned and have had great results but are still looking into reducing the latency as much as possible, which is why we are trying to get jamulus to work with 32 or lower frame size.
We have already achieved 32 frame audio buffer size for client side using JACK but are having trouble to get the same at server side.

What changes do I need to make to get server side to run with 32 frame audio buffer size as well?
Where do these changes need to be?

**Reply by ann0see** [reply]
I'm not sure, but probably it's somewhere in util.h/.cpp. You can also look for "64" in the changelog and the respective commit for 64 size support. 

Edit: it's Version 3.4.7

**Reply by Meemansa-Mishra** [reply]
@ann0see 
Thanks I will start checking it out.

Also I found something in comments in the code(util.cpp), could you let me know if I will encounter any trouble reducing frame size due to this?
![Screenshot from 2022-09-12 16-17-37](https://user-images.githubusercontent.com/36433747/189635335-e3246e8c-6442-40dd-b117-a9a985c09708.png)

**Reply by ann0see** [reply]
Sorry, but I don't know. I'd just try it, look for the commit which introduced 64 samples and try to understand what was changed.

**Reply by pljones** [reply]
The issue you'll have is that the highest resolution timer available on certain platforms is 1ms.

32 samples at 48 thousand samples per second means you don't get a "tick" from the timer for each 32 samples.  So you'll **miss** exact timing for your 32 samples and introduce jitter:

```
32 sample frame size at 48KHz sampling rate with 1ms timer resolution:
actual intervals: 0.0   0.667   1.333   2.000   2.667   3.333   4.000
quantised to 2ms: 0     1       1       2       2       3       4
```
So on "1", you get to read the first _two_ buffers; on "2" you get the next two buffers; but on "3", you only get _one_ buffer.

**Comment by ann0see** [comment]
Hi @Meemansa-Mishra 
You want to lower the latency, I suppose? There are multiple places where you can tweak the latency; especially your network and audio hardware are bottlenecks. Do you really suffer from high latency (Latency >= 100ms) or do you just want to achieve the lowest possible latency?


Note: You can use JACK to lower audio buffer size, but the question is if that introduces side-effects: https://github.com/jamulussoftware/jamulus/discussions/2844#discussioncomment-3610177

**Reply by Meemansa-Mishra** [reply]
Hi @ann0see 
I don't suffer from high latency but I am trying to achieve the lowest possible latency and would like to go below 64 frame size.(I have 48 or 32 in mind).We have already achieved 32 frame audio buffer size for client side using JACK but are having trouble to get the same at server side.

What changes do I need to make to get server side to run with 32 frame audio buffer size as well?
Where do these changes need to be?

**Reply by pljones** [reply]
They'd need to be similar to the way the 64 sample and 128 sample frame sizes are handled by server.cpp.

**Reply by parveensachdeva** [reply]
hi @pljones @ann0see 
Can you please tell where and how network buffer is getting set ? what it is by default ?
Appriciate your help.

**Reply by pljones** [reply]
I'm not sure - try `src/socket.cpp`.  I think the network frame size is roughly speaking the network packet size plus a Jamulus small header, plus the UDP and IP protocol headers.  The buffer, as such, will be managed by the network code.

There's the jitter buffer for each channel, which is managed in `src/channel.cpp`, too.  You'll see `iAudioFrameSizeSamples ( DOUBLE_SYSTEM_FRAME_SIZE_SAMPLES ),` very early on in the file.  That setting was introduced when support for 64 sample frames was added.

**Comment by pljones** [comment]
And remember, like @DavidSavinkoff points out, the difference between 64 samples and 32 samples is 6.75 inches.
